File: solutions/187/main.py
from typing import * 

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        L = 10 
        idx = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
        
        n = len(s)
        if n < L: 
            return []
        
        # total 8 x 4 = 32 bits 
        mask = 0x000FFFFF
        state = 0 
        ans = []
        memo = {}
        
        update_state = lambda state, ch: ((state << 2) | idx[ch]) & mask 
        
        # NOTE, L-1, not L!!!
        # only put the first 9 characters into state 
        for i in range(L-1): 
            state = update_state(state, s[i]) 
        
        for i in range(n-L+1): 
            state = update_state(state, s[i+L-1]) 
            memo[state] = memo.get(state, 0) + 1
            if memo[state] == 2: 
                # NOTE, do not add the same substr
                # only add when count == 2 
                ans.append(s[i:i+L])
              
        return ans 

if __name__ == "__main__": 
    s = Solution()
    
    assert s.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT") == ["AAAAACCCCC","CCCCCAAAAA"]
        
        
        

File: solutions/187/main.cpp
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        int n = s.length(); 
        if (n <= 10) return {}; 

        int state = 0; 
        unordered_map<char, int> idx; 
        idx['A'] = 0; idx['C'] = 1; idx['G'] = 2; idx['T'] = 3;

        vector<string> ans; 
        // 0x is base 16 
        // need 2*10 = 20 bits to record each substring 
        // 20 bits are FFFFF in 0x 
        int mask = 0x000FFFFF, record[mask+1]; 
        memset(record, 0, sizeof record); 

        // process the first 10 chars in s 
        for (int i = 0; i < 10; i++) {
            state = state << 2 | idx[s[i]]; 
        }
        ++record[state]; 

        for (int i = 10; i < n; ++i) {
            state = (state << 2 | idx[s[i]]) & mask; 
            if (record[state] == 1) {
                ans.emplace_back(s.substr(i-9, 10)); 
            } 
            ++record[state]; 
        }

        return ans; 
    }
};